### PR TITLE
[mlir][affine] implement `promoteIfSingleIteration` for `AffineForOp`

### PR DESCRIPTION
**Pull request by makslevental**
Recently someone on discord asked if `AffineForOp` could be refactored to just implement the `promoteIfSingleIteration` interface method on `LoopLikeOpInterface` (rather than have a dangling method). Went more smoothly than I thought so here it is.

### CODE REVIEW

File: mlir/lib/Dialect/Affine/Analysis/LoopAnalysis.cpp [hunk 1]
```diff
@@ -107,6 +107,20 @@ std::optional<uint64_t> mlir::affine::getConstantTripCount(AffineForOp forOp) {
   return tripCount;
 }
 
+void mlir::affine::replaceIterArgsAndYieldResults(AffineForOp forOp) {
+  // Replace uses of iter arguments with iter operands (initial values).
+  auto iterOperands = forOp.getInits();
+  auto iterArgs = forOp.getRegionIterArgs();
+  for (auto e : llvm::zip(iterOperands, iterArgs))
+    std::get<1>(e).replaceAllUsesWith(std::get<0>(e));
```
Comment (kuhar) [inline mlir/lib/Dialect/Affine/Analysis/LoopAnalysis.cpp:115]:
```suggestion
  for (auto [operand, arg] : llvm::zip(iterOperands, iterArgs))
    arg.replaceAllUsesWith(operand);
```

File: mlir/lib/Dialect/Affine/Analysis/LoopAnalysis.cpp [hunk 2]
```diff
@@ -107,6 +107,20 @@ std::optional<uint64_t> mlir::affine::getConstantTripCount(AffineForOp forOp) {
...
+
+  // Replace uses of loop results with the values yielded by the loop.
+  auto outerResults = forOp.getResults();
+  auto innerResults = forOp.getBody()->getTerminator()->getOperands();
+  for (auto e : llvm::zip(outerResults, innerResults))
+    std::get<0>(e).replaceAllUsesWith(std::get<1>(e));
```
Comment (kuhar) [inline mlir/lib/Dialect/Affine/Analysis/LoopAnalysis.cpp:121]: Same here

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp [hunk 3]
```diff
@@ -2440,6 +2442,53 @@ std::optional<OpFoldResult> AffineForOp::getSingleUpperBound() {
...
+  if (!tripCount || *tripCount != 1)
+    return failure();
+
+  // TODO: extend this for arbitrary affine bounds.
+  if (forOp.getLowerBoundMap().getNumResults() != 1)
+    return failure();
+
+  // Replaces all IV uses to its single iteration value.
+  auto iv = forOp.getInductionVar();
+  auto *parentBlock = forOp->getBlock();
```
Comment (kuhar) [inline mlir/lib/Dialect/Affine/IR/AffineOps.cpp:2459]: Same nit about types

**Review by kuhar — COMMENTED** [review]
Just some cosmetics that stood out to me. Would be nice to fix, but I'm fine with leaving them as-is for the exiting code that has been moved across files if you really don't want to change this.

**Comment by makslevental** [comment]
> Just some cosmetics that stood out to me. Would be nice to fix, but I'm fine with leaving them as-is for the exiting code that has been moved across files if you really don't want to change this.

I'm all for sprucing things up while I'm here.

**Review by kuhar — APPROVED** [review]
Thanks

**Pull request merged**